executor/aggfuncs/func_count_distinct.go

```diff
@@ -433,3 +435,386 @@ func appendJSON(encodedBytes, _ []byte, val json.BinaryJSON) []byte {
+	buf        []approxCountDistinctHashValue
+}
+
+// NewPartialResult4ApproxCountDistinct builds a partial result for agg function ApproxCountDistinct.
+func NewPartialResult4ApproxCountDistinct() *partialResult4ApproxCountDistinct {
+	p := &partialResult4ApproxCountDistinct{}
+	p.reset()
+	return p
+}
+
+func (p *partialResult4ApproxCountDistinct) InsertHash64(x uint64) {
+	// no need to rehash, just cast into uint32
+	p.insertHash(approxCountDistinctHashValue(x))
+}
+
+func (p *partialResult4ApproxCountDistinct) alloc(newSizeDegree uint8) {
+	p.size = 0
+	p.skipDegree = 0
+	p.hasZero = false
+	p.buf = make([]approxCountDistinctHashValue, uint32(1)<<newSizeDegree)
+	p.sizeDegree = newSizeDegree
+}
+
+func (p *partialResult4ApproxCountDistinct) reset() {
+	p.alloc(uniquesHashSetInitialSizeDegree)
+}
+
+func max(a, b uint8) uint8 {
+	if a > b {
+		return a
+	}
+
+	return b
+}
+
+func (p *partialResult4ApproxCountDistinct) bufSize() uint32 {
+	return uint32(1) << p.sizeDegree
+}
+
+func (p *partialResult4ApproxCountDistinct) mask() uint32 {
+	return p.bufSize() - 1
+}
+
+func (p *partialResult4ApproxCountDistinct) place(x approxCountDistinctHashValue) uint32 {
+	return uint32(x>>uniquesHashBitsForSkip) & p.mask()
+}
+
+// Increase the size of the buffer 2 times or up to new size degree.
+func (p *partialResult4ApproxCountDistinct) resize(newSizeDegree uint8) {
+	oldSize := p.bufSize()
+	oldBuf := p.buf
+
+	if 0 == newSizeDegree {
+		newSizeDegree = p.sizeDegree + 1
+	}
+
+	p.buf = make([]approxCountDistinctHashValue, uint32(1)<<newSizeDegree)
+	p.sizeDegree = newSizeDegree
+
+	// Move some items to new locations.
+	for i := uint32(0); i < oldSize; i++ {
+		x := oldBuf[i]
+		if x != 0 {
+			p.reinsertImpl(x)
+		}
+	}
+}
+
+func (p *partialResult4ApproxCountDistinct) readAndMerge(rb []byte) error {
+	rhsSkipDegree := rb[0]
+	rb = rb[1:]
+
+	if rhsSkipDegree > p.skipDegree {
+		p.skipDegree = rhsSkipDegree
+		p.rehash()
+	}
+
+	rb, rhsSize, err := codec.DecodeUvarint(rb)
+
+	if err != nil {
+		return err
+	}
+
+	if rhsSize > uint64(uniquesHashMaxSize) {
+		return errors.New("Cannot read partialResult4ApproxCountDistinct: too large size degree")
+	}
+
+	if p.bufSize() < uint32(rhsSize) {
+		newSizeDegree := max(uniquesHashSetInitialSizeDegree, uint8(math.Log2(float64(rhsSize-1)))+2)
+		p.resize(newSizeDegree)
+	}
+
+	for i := uint32(0); i < uint32(rhsSize); i++ {
+		x := *(*approxCountDistinctHashValue)(unsafe.Pointer(&rb[0]))
+		rb = rb[4:]
+		p.insertHash(x)
+	}
+
+	return err
+}
+
+func (p *partialResult4ApproxCountDistinct) fixedSize() uint64 {
+	if 0 == p.skipDegree {
+		return uint64(p.size)
+	}
+
+	res := uint64(p.size) * (uint64(1) << p.skipDegree)
+
+	// Pseudo-random remainder.
+	res += intHash64(uint64(p.size)) & ((uint64(1) << p.skipDegree) - 1)
+
+	// Correct system errors due to collisions during hashing in uint32.
+	p32 := uint64(1) << 32
+	fixedRes := math.Round(float64(p32) * (math.Log(float64(p32)) - math.Log(float64(p32-res))))
```

Comment:
What if `res` exceeds `p32`? It looks wired.

Reply:
When res exceeds p32, fixedRes will become a very large number like 1.8 * 10^18.
This algorithm may lose accuracy when number of distinct element is more than 2^32.

Reply:
Yeah, can we add a code comment for this phenomenon?